### PR TITLE
DE461003: Fixing policies.yml serialization and deserialization

### PR DESCRIPTION
**Pull request by nikhilahuja**
Fixes #
- policies.yml will now have only direct dependencies under the `usedEntities`
- Duirng `build` task we will recursively identify all the dependencies for the bundle.

### CODE REVIEW

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java [hunk 1]
```diff
@@ -158,75 +158,100 @@ private void removeEntitiesForDeleteBundle(final List<Entity> entities) {
...
+
+        Map<String, Policy> annotatedPolicyMap = annotatedBundle.getEntities(Policy.class);
+        annotatedPolicyMap.put(policy.getPath(), policy);
+
+        Set<Dependency> dependencies = policy.getUsedEntities();
+        if (dependencies != null) {
+            for (Dependency dependency : dependencies) {
+                switch (dependency.getType()) {
+                    case EntityTypes.POLICY_TYPE:
+                        Policy dependentPolicy = rawBundle.getPolicies().get(dependency.getName());
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:197]: rawBundle Policy map will have key as path, but dependency has name

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java [hunk 2]
```diff
@@ -158,75 +158,100 @@ private void removeEntitiesForDeleteBundle(final List<Entity> entities) {
...
-    private void insertGatewayEntity(Map entityMapToUpdate, String key, GatewayEntity gatewayEntity,
-                                     boolean excludeReusable, List<GatewayEntity> excludedReusableEntities) {
-        if (excludeReusable && gatewayEntity instanceof AnnotableEntity && ((AnnotableEntity) gatewayEntity).isReusable()) {
-            excludedReusableEntities.add(gatewayEntity);
-            return; // Return without inserting as its reusable entity
+    private void loadEncassDependencies(Encass encass, AnnotatedBundle annotatedBundle, Bundle rawBundle,
+                                        boolean excludeReusable) {
+        if (!excludeGatewayEntity(Encass.class, encass, annotatedBundle, excludeReusable)) {
+            annotatedBundle.getEncasses().put(encass.getName(), encass);
+            loadPolicyDependenciesByPolicyName(encass.getPolicy(), annotatedBundle, rawBundle, excludeReusable);
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:223]: encass.getPolicy gives full path or only name ?

**Comment by uppoju** [comment]
Please fix functional test cases in Dev test repo for export project

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java (hunk 2 above)
Comment (nikhilahuja) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:223]: Encass contains full path.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java (hunk 1 above)
Comment (nikhilahuja) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java]: Fixed it!

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java [hunk 3]
```diff
@@ -158,75 +158,97 @@ private void removeEntitiesForDeleteBundle(final List<Entity> entities) {
...
      */
-    private void insertGatewayEntity(Map entityMapToUpdate, String key, GatewayEntity gatewayEntity,
-                                     boolean excludeReusable, List<GatewayEntity> excludedReusableEntities) {
-        if (excludeReusable && gatewayEntity instanceof AnnotableEntity && ((AnnotableEntity) gatewayEntity).isReusable()) {
-            excludedReusableEntities.add(gatewayEntity);
-            return; // Return without inserting as its reusable entity
+    private void loadEncassDependencies(Encass encass, AnnotatedBundle annotatedBundle, Bundle rawBundle,
+                                        boolean excludeReusable) {
+        if (encass != null && !excludeGatewayEntity(Encass.class, encass, annotatedBundle, excludeReusable)) {
+            annotatedBundle.getEncasses().put(encass.getName(), encass);
```
Comment (RaghuReddy) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:219]: just a check, encass.getName() is giving you the name or null?

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java [hunk 4]
```diff
@@ -236,20 +258,56 @@ private void removeExcludedReusablePolicy(Policy policy, AnnotatedBundle annotat
...
+     *
+     * @param entityType Type of entity class
+     * @param gatewayEntity Gateway entity to be checked
+     * @param annotatedBundle Annotated Bundle for which bundle is being created.
+     * @param excludeReusable Exclude loading Reusable entities as the dependency
+     * @return TRUE if the Gateway entity needs to be excluded
+     */
+    private boolean excludeGatewayEntity(Class<? extends GatewayEntity> entityType, GatewayEntity gatewayEntity,
+                                         AnnotatedBundle annotatedBundle, boolean excludeReusable) {
+        return excludeReusableEntity(gatewayEntity, excludeReusable)
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:276]: Can you please merge excludeReusableEntity and excludePolicy methods into here.

And rewrite the logic like below:
if (<condition1> {
return true;
}

if (<condition2>) {
return true;
}

....

return false;

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java (hunk 3 above)
Comment (nikhilahuja) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:219]: @RaghuReddy `getName()` should not give null as we set it in `Encass.postLoad()`

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java (hunk 4 above)
Comment (nikhilahuja) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java]: @graju256 Done!

**Pull request merged**